### ISSUE TITLE
Zero lines covered in Cobertura yields confusing error messaage

### ISSUE THREAD

**Issue by tillig** [microsoft/azure-pipelines-tasks#12634]
## Required Information

- **Type**: Bug
- **Task Name**: `PublishCodeCoverageResults`

## Environment

- Server - Azure Pipelines
- Agent - Hosted / Ubuntu

## Issue Description

I'm setting up a build for a brand new .NET Core project. There are skeleton projects in there that have not been totally filled out yet, but it's enough to compile and run Hello World. As part of that, I'm setting up coverage reporting using Coverlet and Cobertura output.

When I run tests with coverage, I do get a valid XML coverage report coming out. I then run that through [the report generator extension](https://marketplace.visualstudio.com/items?itemName=Palmmedia.reportgenerator) and, finally, use the aggregated report to `PublishCodeCoverageResults`.

However, while there _is a report_ there is not currently any actual _line coverage_ because the tests aren't quite there yet. This yields a very confusing error message withe a debug log like this:

```text
##[debug]findPath: '/home/vsts/work/1/a/log/coverage/Cobertura.xml'
##[debug]statOnly: 'true'
##[debug]found 1 paths
##[debug]applying include pattern
##[debug]1 matches
##[debug]1 final results
##[debug]Found 1 result(s) matching pattern: /home/vsts/work/1/a/log/coverage/Cobertura.xml
##[debug]Resolved summary file: /home/vsts/work/1/a/log/coverage/Cobertura.xml
##[debug]Reading XML file: /home/vsts/work/1/a/log/coverage/Cobertura.xml
##[debug]Converting XML file to JSON
##[debug]task result: Failed
##[error]No code coverage results were found to publish.
```

There _are coverage results_ and they _could be published_ but it's saying nothing was found.

The error is [here](https://github.com/microsoft/azure-pipelines-tasks/blob/5d99f827ece0b04fb09f7b290a00e7fc64102f7a/Tasks/Common/codecoverage-tools/codecoverageutilities.ts#L16):

```typescript
else if (codeCoverageTool.toLowerCase() === 'cobertura' && resp.coverage) {
  let lines_covered: number = Number(resp.coverage.$["lines-covered"]);
  if (lines_covered && lines_covered > 0) {
    return Q.resolve(false);
  }
}
```

Basically, it says if there are zero lines covered, return that there was no coverage found. Way up the stack, that yields a very generic "no results" message. The message shown in the logs says that there weren't _any_ results; not that the results showed no coverage.

While this seems like a minor semantic issue, I honestly just sank like a day into debugging deep into the task trying to figure out whether there was a problem with the report format, or the report generator plugin, or whether maybe the `xml2js` module couldn't read the file, or something else. It might be good to have a debug message in there to say something about how there _are results_ it's just that there's nothing of value to report. It'd have saved a lot of time.

**Comment by viktors-telle** [comment]
I agree that the error message should be improved. I've spent a lot of hours trying to debug the issue because of the generic error message.

**Comment by tillig** [comment]
The error message here still needs improvement.

**Comment by tillig** [comment]
[The location of the original issue has changed](https://github.com/microsoft/azure-pipelines-tasks/blob/7203e1c1f3c9addc9f42920789bc1b3a99bc90ef/common-npm-packages/codecoverage-tools/codecoverageutilities.ts#L16) and the issue is still present in that code. However, it appears to only affect [`PublishCodeCoverageResults@1`](https://github.com/microsoft/azure-pipelines-tasks/blob/7203e1c1f3c9addc9f42920789bc1b3a99bc90ef/Tasks/PublishCodeCoverageResultsV1/publishcodecoverageresults.ts#L40) while there's a whole [`PublishCodeCoverageResults@2` task](https://github.com/microsoft/azure-pipelines-tasks/blob/7203e1c1f3c9addc9f42920789bc1b3a99bc90ef/Tasks/PublishCodeCoverageResultsV2/publishcodecoverageresults.ts#L1) that doesn't use this utility.

However, [the documentation doesn't mention the @2 version of the task](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/test/publish-code-coverage-results?view=azure-devops) which is confusing.

Is it just a documentation issue?

**Comment by tillig** [comment]
[I filed an issue to ask if this is a missing doc problem](https://github.com/MicrosoftDocs/azure-devops-docs/issues/11913) but it'd be nice to hear from the team on this.
